tracer/src/Datadog.Trace/Util/Http/QueryStringObfuscator.cs

```diff
@@ -0,0 +1,94 @@
+    {
+        /// <summary>
+        /// Default obfuscation query string regex if none specified via env DD_OBFUSCATION_QUERY_STRING_REGEXP
+        /// </summary>
+        public const string DefaultObfuscationQueryStringRegex = @"((?i)(?:p(?:ass)?w(?:or)?d|pass(?:_?phrase)?|secret|(?:api_?|private_?|public_?|access_?|secret_?)key(?:_?id)?|token|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)(?:(?:\s|%20)*(?:=|%3D)[^&]+|(?:""|%22)(?:\s|%20)*(?::|%3A)(?:\s|%20)*(?:""|%22)(?:%2[^2]|%[^2]|[^""%])+(?:""|%22))|bearer(?:\s|%20)+[a-z0-9\._\-]|token(?::|%3A)[a-z0-9]{13}|gh[opsu]_[0-9a-zA-Z]{36}|ey[I-L](?:[\w=-]|%3D)+\.ey[I-L](?:[\w=-]|%3D)+(?:\.(?:[\w.+\/=-]|%3D|%2F|%2B)+)?|[\-]{5}BEGIN(?:[a-z\s]|%20)+PRIVATE(?:\s|%20)KEY[\-]{5}[^\-]+[\-]{5}END(?:[a-z\s]|%20)+PRIVATE(?:\s|%20)KEY|ssh-rsa(?:\s|%20)*(?:[a-z0-9\/\.+]|%2F|%5C|%2B){100,})";
+
+        private static readonly IDatadogLogger _log = DatadogLogging.GetLoggerFor(typeof(QueryStringObfuscator));
+        private static QueryStringObfuscator _instance;
+        private static bool _globalInstanceInitialized;
+        private static object _globalInstanceLock = new();
+        private readonly Obfuscator _obfuscator;
+
+        private QueryStringObfuscator(double timeout, string pattern = null)
+        {
+            pattern ??= Tracer.Instance.Settings.ObfuscationQueryStringRegex;
+            _obfuscator = new(TimeSpan.FromMilliseconds(timeout), pattern);
+        }
+
+        internal string Obfuscate(string queryString) => _obfuscator.Obfuscate(queryString);
+
+        /// <summary>
+        /// Gets or sets the global <see cref="QueryStringObfuscator"/> instance.
+        /// </summary>
+        public static QueryStringObfuscator Instance(double timeout, string pattern = null) => LazyInitializer.EnsureInitialized(ref _instance, ref _globalInstanceInitialized, ref _globalInstanceLock, () => new(timeout, pattern));
+
+        internal class Obfuscator
+        {
+            private const string ReplacementString = "<redacted>";
+            private readonly Regex _regex;
+            private readonly bool _disabled;
+            private readonly TimeSpan _timeout;
+
+            internal Obfuscator(TimeSpan timeout, string pattern = null)
+            {
+                _timeout = timeout;
+                if (string.IsNullOrEmpty(pattern))
+                {
+                    _disabled = true;
+                }
+                else
+                {
+                    _regex = new(pattern, RegexOptions.IgnoreCase | RegexOptions.Compiled, _timeout);
+                }
+            }
+
+            internal string Obfuscate(string queryString)
+            {
+                if (_disabled || string.IsNullOrEmpty(queryString))
+                {
+                    return queryString;
+                }
+
+                try
+                {
+                    queryString = queryString.Substring(0, Math.Min(queryString.Length, 200));
```

Comment:
I assume you do this for perf reasons. Why have you chosen this number? Is security ok with not obfuscating everything?

Reply:
yes it's better working and reporting a long enough query string (200) than timing out on what could likely be an injection